File: go/test/post_title_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/healthcare-gov-content-sdk/go"
	"github.com/voxgig-sdk/healthcare-gov-content-sdk/go/core"

	vs "github.com/voxgig-sdk/healthcare-gov-content-sdk/go/utility/struct"
)

func TestPostTitleEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.PostTitle(nil)
		if ent == nil {
			t.Fatal("expected non-nil PostTitleEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := post_titleBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "post_title." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set HEALTHCAREGOVCONTENT_TEST_POST_TITLE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		postTitleRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.post_title", setup.data)))
		var postTitleRef01Data map[string]any
		if len(postTitleRef01DataRaw) > 0 {
			postTitleRef01Data = core.ToMapAny(postTitleRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = postTitleRef01Data

		// LIST
		postTitleRef01Ent := client.PostTitle(nil)
		postTitleRef01Match := map[string]any{
			"post_title": setup.idmap["post_title01"],
		}

		postTitleRef01ListResult, err := postTitleRef01Ent.List(postTitleRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, postTitleRef01ListOk := postTitleRef01ListResult.([]any)
		if !postTitleRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", postTitleRef01ListResult)
		}

	})
}

func post_titleBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "post_title", "PostTitleTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read post_title test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse post_title test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"post_title01", "post_title02", "post_title03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("HEALTHCAREGOVCONTENT_TEST_POST_TITLE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"HEALTHCAREGOVCONTENT_TEST_POST_TITLE_ENTID": idmap,
		"HEALTHCAREGOVCONTENT_TEST_LIVE":      "FALSE",
		"HEALTHCAREGOVCONTENT_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["HEALTHCAREGOVCONTENT_TEST_POST_TITLE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["HEALTHCAREGOVCONTENT_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewHealthcareGovContentSDK(core.ToMapAny(mergedOpts))
	}

	live := env["HEALTHCAREGOVCONTENT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["HEALTHCAREGOVCONTENT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/post_title_entity_test.rb
# PostTitle entity test

require "minitest/autorun"
require "json"
require_relative "../HealthcareGovContent_sdk"
require_relative "runner"

class PostTitleEntityTest < Minitest::Test
  def test_create_instance
    testsdk = HealthcareGovContentSDK.test(nil, nil)
    ent = testsdk.PostTitle(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = post_title_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "post_title." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set HEALTHCAREGOVCONTENT_TEST_POST_TITLE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    post_title_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.post_title")))
    post_title_ref01_data = nil
    if post_title_ref01_data_raw.length > 0
      post_title_ref01_data = Helpers.to_map(post_title_ref01_data_raw[0][1])
    end

    # LIST
    post_title_ref01_ent = client.PostTitle(nil)
    post_title_ref01_match = {
      "post_title" => setup[:idmap]["post_title01"],
    }

    post_title_ref01_list_result, err = post_title_ref01_ent.list(post_title_ref01_match, nil)
    assert_nil err
    assert post_title_ref01_list_result.is_a?(Array)

  end
end

def post_title_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "post_title", "PostTitleTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = HealthcareGovContentSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["post_title01", "post_title02", "post_title03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["HEALTHCAREGOVCONTENT_TEST_POST_TITLE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "HEALTHCAREGOVCONTENT_TEST_POST_TITLE_ENTID" => idmap,
    "HEALTHCAREGOVCONTENT_TEST_LIVE" => "FALSE",
    "HEALTHCAREGOVCONTENT_TEST_EXPLAIN" => "FALSE",
    "HEALTHCAREGOVCONTENT_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["HEALTHCAREGOVCONTENT_TEST_POST_TITLE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["HEALTHCAREGOVCONTENT_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["HEALTHCAREGOVCONTENT_APIKEY"],
      },
      extra || {},
    ])
    client = HealthcareGovContentSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["HEALTHCAREGOVCONTENT_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["HEALTHCAREGOVCONTENT_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
